Legacy-to-PSA transition guide

gilles-peskine-arm opened the pull request:
A new document explains how to migrate application code from the legacy crypto API to the PSA API.

I initially thought this would go in the knowledge base, but we're going to have different versions in the 3.x series and in the 4.x series, so upon reflection it should be versioned. The document mostly focuses on what is in Mbed TLS today, but includes a few notes about upcoming changes.

As I was writing this document, some gaps between the two APIs became evident. I created a few issues for these gaps, with a link in the document.

## Reviewers

The top-level sections are largely independent, so we are splitting the review load. For each section, we will have at least one person reviewing for clarity and one for correctness.

* Clarity (including completeness of coverage): is the document comprehensible by people who aren't familiar with the PSA API, and maybe not with the legacy API either? The document isn't meant to be self-contained: readers are supposed to refer to the Mbed TLS reference documentation. The goal of this document is to allow readers to determine which functions to call and how to combine them. Note that this includes a completeness aspect: are all the elements of the legacy API discussed?
* Correctness (including completeness of purpose): is this document making incorrect claims? Are there major use cases of the legacy API that are not covered but should be?

I recommend that all reviewers read “Introduction” and at least skim “General considerations” to get an idea of the goals of the document.

| Sections | Length | Reviewers for correctness | Reviewers for clarity |
|--|--|--|--|
| Introduction | S | Ronald :white_check_mark: | @tom-daubney-arm :white_check_mark: |
| General considerations | S | Ronald :white_check_mark: | @tom-daubney-arm :white_check_mark: |
| Summary of API modules | S | Ronald :white_check_mark: | @tom-daubney-arm :white_check_mark: |
| Compile-time configuration | M | Ronald :white_check_mark: | @tom-daubney-arm :white_check_mark: |
| Miscellaneous support modules | S | Ronald :white_check_mark: | @tom-daubney-arm :white_check_mark: |
| Symmetric encryption | L | @davidhorstmann-arm :white_check_mark: |@Ryan-Everett-arm :white_check_mark:  |
| Hashes and MAC | L |  @gowthamsk-arm:white_check_mark: | @davidhorstmann-arm :white_check_mark: |
| Key derivation | S | @valeriosetti:white_check_mark:  | @tgonzalezorlandoarm:white_check_mark: |
| Random generation | S | @valeriosetti:white_check_mark: | @tgonzalezorlandoarm:white_check_mark: |
| Asymmetric cryptography | XL | Manuel :white_check_mark: | @tom-daubney-arm :white_check_mark: |

## PR checklist

- [x] **changelog** no (doc only)
- [x] **backport** no (no point in a 2.28 migration guide)
- [x] **tests** no (doc only)

mpg commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1064 @@
+
+To make the PSA API available, make sure that the configuration option [`MBEDTLS_PSA_CRYPTO_C`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/file/mbedtls__config_8h/#c.MBEDTLS_PSA_CRYPTO_C) is enabled (it is enabled in the default configuration).
+
+You should probably enable [`MBEDTLS_USE_PSA_CRYPTO`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/file/mbedtls__config_8h/#mbedtls__config_8h_1a70fd7b97d5f11170546583f2095942a6) as well (it is disabled by default). This option causes the PK, X.509 and TLS modules to use PSA crypto under the hood. Some functions that facilitate the transition are only available when `MBEDTLS_US ...
```
Should we mention here that enabling `USE_PSA_CRYPTO` also means you promise to call `psa_crypto_init()` before doing PK/X.509/TLS operations?

gilles-peskine-arm replied:
I'm hesitating between being precise and not drowning the reader. This is the must-read section, so I lean towards not drowning the header. Keep it simple (“Before any cryptographic operation”), and let the user read the documentation for the fine print (that it's only before doing any cryptographic operation via a PSA API, or via a non-PSA API that calls a PSA API and what that means for PK/X509/TLS). So I don't intend a change here.

tgonzalezorlandoarm replied:
What about putting an example on one function that "facilitates the transition" ? I'm not sure I understand exactly what that means otherwise, that's why I ask

gilles-peskine-arm replied:
Because this is an introductory section, I'd rather stick to general principles and not include examples. The examples are in sections about specific topics (cipher, ECC, …), but there I don't discuss which functions need exactly which compilation options, because that's something users can very easily find in the documentation.

Would it help if the document explained what kind of things such functions might do?

> Some functions that facilitate the transition (e.g. conversions between metadata encodings or between key representations) are only available (…)

tgonzalezorlandoarm replied:
Yes, I think that would help a lot. Thanks!

tgonzalezorlandoarm replied:
Ack

mpg commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1064 @@
+    * `PSA_KEY_TYPE_xxx_IMPORT` enables support for `psa_import_key` to import a key of that type.
+    * `PSA_KEY_TYPE_xxx_GENERATE` enables support for `psa_generate_key` to randomly generate a key of that type.
+    * `PSA_KEY_TYPE_xxx_DERIVE` enables support for `psa_key_derivation_output_key` to deterministically derive a key of that type.
+    * `PSA_KEY_TYPE_xxx_EXPORT` enables support for `psa_export_key` to export a key of that type.
```
Stupid question about ordering: doesn't seem to be alphabetical, and export not grouped with import either, so is there a logic I'm missing?

gilles-peskine-arm replied:
1. USE because it's the “most basic” one.
2. Creation methods.
3. What you can do after creating the key.

I suppose it would make more sense to move USE down.

mpg replied:
I think the two orderings that make the most sense to me are either: use, export, import, generate, derive (roughly "most basic" to least), or import, generate, derive, use, export (as you said, create then use). Slight preference for the first one, keeping USE first, but OK with anything.

gilles-peskine-arm replied:
I'm changing `USE` to `BASIC` per https://github.com/Mbed-TLS/mbedtls/issues/7439#issuecomment-1592673401. I think with the changed name and intuition, the current order now makes sense.

mpg commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1064 @@
+* To make `PSA_KEY_TYPE_ttt` available, enable `PSA_WANT_KEY_TYPE_ttt`.
+
+    As an exception, starting in Mbed TLS 3.5.0, for key pair types, the feature selection is more fine-grained, with an additional suffix:
+    * `PSA_KEY_TYPE_xxx_USE` enables support for operations with a key of that type (for enabled algorithms). This is automatically enabled if any key creation method (`IMPORT`, `GENERATE` or `DERIVE`) is enabled.
```
This should be `PSA_WANT_` right?

gilles-peskine-arm commented:
I have now covered `ecp.h` and have no further changes planned for this first edition of the document, except for handling review comments.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -800,16 +800,30 @@ psa_import_key(&attributes, buf + sizeof(buf) - length, length, &key_id);
+    mbedtls_mpi_init(&d);
+    mbedtls_ecp_export(ec, &grp, &d, &Q);
+    size_t bits;
+    curve = mbedtls_ecc_group_to_psa(grp.id, &bits);
```
Turns out we did have the ECP→PSA bridge function here, since the beginning. On the other hand the way to extract metadata from the legacy object is ridiculous because we haven't finished 3.0 properly.

gilles-peskine-arm commented:
I'm raising the priority to very-high because the detailed investigation for the legacy-to-PSA API design depends on this and this investigation is overdue.

mpg commented:
> I have now covered `ecp.h` and have no further changes planned for this first edition of the document, except for handling review comments as they come.

The PR's description still says "missing most of ecp, ecdh and dhm. I'll finish when I have time" - I assume that part's out of date then?

gilles-peskine-arm commented:
@mpg Yes, I had just forgotten to remove that sentence from the PR description.

gilles-peskine-arm commented on docs/psa-transition.md:
Well, when you're writing application code, you decide what size to use for the output buffer. And these macros give you a way to make this decision.

Strictly speaking, the macros don't give the required size of the output buffer: in some cases this is an overapproximation (for example, with block ciphers, you can get away with a smaller size for `psa_cipher__update` if you call it on a whole number of blocks each time, whereas `PSA_CIPHER_ENCRYPT_UPDATE_OUTPUT_SIZE` has to cater for the worst case with unaligned chunks.)

Do you think “to determine a sufficient size for the output buffer” would be clearer? It's wordier and I don't feel that it conveys any extra useful information.

Ryan-Everett-arm replied:
Upon reading the current phrasing my first impression was that the size of the output buffer is fixed - of course just clicking on either link will inform a reader that this is not the case, so perhaps I am being too nit-picky. I do think "to determine a sufficient size .." is clearer as it doesn't give this impression.

tgonzalezorlandoarm commented on docs/psa-transition.md:
> `INIT`

Maybe link to https://mbed-tls.readthedocs.io/projects/api/en/development/api/group/group__key__derivation/#group__key__derivation_1ga5bdec40d201551558579a635520ae85d ?

gilles-peskine-arm replied:
I don't have an automated way to go from API element name to doxygen link, so I've sometimes skimped on linking API elements. I'll defer that to a polishing follow-up.

tgonzalezorlandoarm commented on docs/psa-transition.md:
> psa_key_derivation_verify_bytes

https://mbed-tls.readthedocs.io/projects/api/en/development/api/group/group__key__derivation/?highlight=psa_key_derivation_verify_bytes#_CPPv431psa_key_derivation_verify_bytesP30psa_key_derivation_operation_tPK7uint8_t6size_t

tgonzalezorlandoarm replied:
Same could happen with the rest?

gilles-peskine-arm replied:
The verify functions didn't exist yet (they were still in development) when I originally wrote this. I've added links for them.

tgonzalezorlandoarm commented on docs/psa-transition.md:
> inteded

nit: intended

tgonzalezorlandoarm replied:
Ack

tgonzalezorlandoarm commented on docs/psa-transition.md:
nit: support

tgonzalezorlandoarm replied:
Ack

tgonzalezorlandoarm commented on docs/psa-transition.md:
> so most applications using PSA crypto do not need the interfaces from `entropy.h`, `ctr_drbg.` and `hmac_drbg.h`

Sorry to be picky, but could we also get a brief explanation or something on when would we need them? Is it only for when we are mixing uses of PSA crypto API and the mbedtls API?

tgonzalezorlandoarm replied:
Ack

tgonzalezorlandoarm reviewed (changes requested):
very small notes and changes

gilles-peskine-arm commented on docs/psa-transition.md:
When a parenthesized expression is exactly one sentence, the final punctuation can be either inside or outside the parentheses, depending on which style guide you follow.

tgonzalezorlandoarm replied:
Ack

tgonzalezorlandoarm commented:
LGTM on my assigned sections

tom-daubney-arm reviewed (changes requested):
Reads very well in general and is clear. Several nits and typos to address.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+
+### Configuration of the PSA subsystem
+
+To make the PSA API available, make sure that the configuration option [`MBEDTLS_PSA_CRYPTO_C`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/file/mbedtls__config_8h/#c.MBEDTLS_PSA_CRYPTO_C) is enabled (it is enabled in the default configuration).
```
Agree with consistency, but I think “in the default configuration” is slightly more precise. (As opposed to e.g. a default that can be changed at runtime — which is [apparently something some people want in some form, for some reason](https://github.com/Mbed-TLS/mbedtls/issues/8567).)

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+
+Before any cryptographic operation, call [`psa_crypto_init`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/group/group__initialization/#group__initialization_1ga2de150803fc2f7dc6101d5af7e921dd9) and check that it succeeds. (A failure indicates an abnormal system state from which most applications cannot recover.)
+
+If you wish to free all resources associated with PSA cryptography, call [`mbedtls_psa_crypto_free`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/file/crypto__extra_8h/#_CPPv423mbedtls_psa_crypto_freev).
```
Adding “the” feels like it's missing a noun, but what? “the PSA cryptography subsystem”? (Not sure I like that word: I think I've already used “subsystem” for various subsets of the library in inconsistent ways.)

tom-daubney-arm replied:
This one is my error. I misread that and upon rereading don't think the "the" is needed as it is written. 

I don't like referring to it as just "PSA cryptography" though: it needs a noun. Agree that subsystem isn't quite right. Maybe something like "the active/current PSA cryptography operations" since "operation" is used above it? Or does that change the semantics too much?

tom-daubney-arm replied:
Fine, we can just leave this one then.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+
+If you wish to free all resources associated with PSA cryptography, call [`mbedtls_psa_crypto_free`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/file/crypto__extra_8h/#_CPPv423mbedtls_psa_crypto_freev).
+
+The PSA subsystem has an internal random generator. As a consequence, you do not need to instantiate one manually (no need to create an `mbedtls_entropy_context` and a `mbedtls_xxx_drbg_context`).
```
Although “random generator” is commonly abbreviated “RNG”, it's not particularly a random _number_ generator. There are more occurrences of “random generator” than “random number generator” in our code base, and in particular PSA documentation almost consistently uses “random generator”. So I prefer to stick to “random generator”.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+The availability of cryptographic mechanisms in the PSA API is based on a systematic pattern:
+
+* To make `PSA_ALG_aaa` available, enable `PSA_WANT_ALG_aaa`.
+  For parametrized algorithms, there is a `PSA_WANT_` symbols both for the main macro and for each argument. For example, to make `PSA_ALG_HMAC(PSA_ALG_SHA_256)` available, enable both `PSA_WANT_ALG_HMAC` and `PSA_WANT_ALG_SHA_256`.
```
I'll fix the plural, but there's a symbol for _each_ of {main macro, first argument, second argument, ...}, not a single symbol for all of them. “Both for …” conveys this more precisely.

tom-daubney-arm replied:
I think the plural threw me off and meant I didn't parse what follows quite correctly. Happy to keep the rest when the plural is fixed.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+
+* To make `PSA_KEY_TYPE_ttt` available, enable `PSA_WANT_KEY_TYPE_ttt`.
+
+    As an exception, starting in Mbed TLS 3.5.0, for key pair types, the feature selection is more fine-grained, with an additional suffix:
```
Hmmm, [debatable](https://english.stackexchange.com/questions/53266/is-the-word-granular-a-synonym-for-the-word-specific/187998#187998). Should it actually be “finer-grained”? That would perhaps be more logical, but seems less idiomatic in 21st century English. For what it's worth: [ngrams](https://books.google.com/ngrams/graph?content=more+fine-grained%2Cfiner-grained%2Cmore+granular&year_start=1800&year_end=2019&corpus=en-2019&smoothing=3).

tom-daubney-arm replied:
Then I bow to your superior knowledge. Sounds funny to me but maybe that's just me. I do not think "finer-grained" sounds good though so stick with what you have.

gilles-peskine-arm replied:
I don't have superior knowledge here. “Granular” doesn't quite feel right to me, and I also think it's less comprehensible to non-native {English,romance} speakers who don't easily recognize that “granular" = “about grain”.

tom-daubney-arm replied:
Well, I was referring to your links which back up your argument. We used to use the word "granular" in physics a fair bit to describe this sort of thing, hence my preference for it I guess. But after reading the links you provided I am satisfied that `fine-grained` is probably the correct choice.

gilles-peskine-arm commented on docs/psa-transition.md:
````diff
@@ -0,0 +1,1293 @@
+#define PSA_WANT_ALG_ECDH
+```
+
+If a mechanism is not enabled by `PSA_WANT_xxx`, Mbed TLS will often not include it, to reduce the size of the compiled library. However, this is not guaranteed: a mechanism that is not explicitly requested can be enabled because it is a dependency of another configuration option, because it is used internally, or because the granularity is not fine enough to distinguish between it and another me ...
````
I've tweaked the first sentence. But I want to keep the repetition of “because”, it really helps to parse the second sentence.

tom-daubney-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+
+### Why change the API?
+
+* Mbed TLS APIs are traditionally very transparent: the caller can access internal fields of operations. This is less true in the 3.x major version than before, but still the case to some extent. This offers applications some flexibility, but it removes flexibility from the implementation. For example, it is hard to support hardware acceleration, because the API constrains how the data must be re ...
```
Because I think it flows and reads better. The "than before" is implied and doesn't need to be written, I don't think.

gilles-peskine-arm commented on docs/psa-transition.md:
Semicolon-hating native speakers vs non-natives desperately looking for better parsing clues than a comma... I'm taking this as “your sentence is too long, don't use a semicolon to get out of it”, and putting a full stop.

tom-daubney-arm replied:
I am not a semicolon hater at all to be fair, I use them myself. I just don't reckon that is quite the right use for one. Two sentences is probably better; the sentence is a bit long.

gilles-peskine-arm commented on docs/psa-transition.md:
Omitting the second “to” doesn't work here because the first “to” doesn't apply to the third sub-clause “for the mask generation …”.

tom-daubney-arm replied:
I don't agree. I think it flows better and I am not sure that it matters that the third sub-clause doesn't have the "to" applying to it. It feels disjointed to me as it is currently expressed. Up to you though. Not going to hold back approval over this.

gilles-peskine-arm replied:
I've tried to reformulate the sentence with the same preposition for all three clauses:

> PSS uses the hash algorithm to hash the message, to create the salted hash, and to generate the mask (as a building block of MGF1).

But even then, I find the sentence clearer when “to” is repeated.

tom-cosgrove-arm replied:
I think it would be considered a matter of personal preference (i.e. the sort of thing covered in a technical writing style guide) as to whether to include the preposition when it's repeated (i.e. the reformulation), but I think it is necessary to include the preposition in the second phrase when it's not the same as the one in the third phrase

tom-daubney-arm replied:
Cool, 2 vs. 1. I bow out :)

gilles-peskine-arm commented on docs/psa-transition.md:
I don't think so: if you're converting between APIs, this isn't your focus. It's only a problem if the old code was already using the risky algorithm.

tom-daubney-arm replied:
Up to you. Whilst I understand that the scope of this document is to provide a guide on how to transition from one API to the other, I do think that there will be users who read this who are just looking for information on the PSA API too. I don't think stressing a potential security concern is a bad thing.

gilles-peskine-arm commented on docs/psa-transition.md:
I'm not sure about this one. Should I replace all 21 occurrences of “flow” in this document by “workflow”?

tom-daubney-arm replied:
I would say workflow is more precise, since you are (as far as I can tell) describing different workflows, so yeah I would replace.

tom-daubney-arm commented on docs/psa-transition.md:
I don't follow.

gilles-peskine-arm replied:
> when reading English prose that my brain automatically groups words in the wrong way, then at some point I notice it doesn't make sense, so I have to go back and try to parse the sentence differently until it makes sense. 

This is a well-known difficulty of English: [garden-path sentences](https://en.wikipedia.org/wiki/Garden-path_sentence). Ambiguous complete sentences are rare, but near-misses are much more common. In most languages, garden-path sentences exist but are much more contrived so they aren't really a concept.

> I guess native speakers have a better ability to guess the correct grouping immediately which is why you don't feel the need for more punctuation.

Native speakers have a quick association from word to meaning, so they often rely on semantic considerations to choose between parsing a word as a verb, noun, adjective, etc. Non-native speakers tend to rely on syntax markers such as word endings (or other characteristics depending on the language) before they perform semantic analysis. English has relatively few syntax markers (no noun declensions, very few verb tenses, practically no number or gender agreement, practically no case markings, frequent null article...). This makes it easier to produce grammatically correct English, but harder to parse idiomatic English.

mpg replied:
Thanks for the reference, never knew it had a name!

tom-daubney-arm commented:
Does a green tick in the reviewers table mean "I have done a review" or "I have done a review and given my approval"? Guessing the latter.

mpg reviewed (commented):
Haven't finished my first pass yet, but I'm uncomfortable having too many pending comments, so I'm releasing them now. I'll continue my review tomorrow.

Note: even though I was reviewing for correctness, most of my comments are about wording - that's mostly because I didn't find a lot of correctness issues :)

gilles-peskine-arm commented on docs/psa-transition.md:
I think it's fine, check the [rendered version](https://github.com/Mbed-TLS/mbedtls/blob/d372da6201a7d7b04e9b595bbe638df9c9b3e7ea/docs/psa-transition.md#asymmetric-cryptography). “and” is in parentheses.

Ryan-Everett-arm commented on docs/psa-transition.md:
There's a double space after "(To encrypt with a specified IV, use the multi-part API described below.)"

gilles-peskine-arm commented on docs/psa-transition.md:
Why? It is supported in the legacy API.

mpg replied:
I read this table as a function "legacy identifier" -> (PSA family, bit size) - where that pair is what you need to identify the curve in PSA. So, for a curve that's not supported I expect (not supported, not supported) as the result.

Anyway, that's minor and not worth arguing over it. If you disagree, just say so and I'll mark this as resolved.

gilles-peskine-arm replied:
Ah, I see, I wasn't paying attention and thought you wanted to remove 224 from the legacy-bit-size column, but there's actually no such column in this table. I'll remove it.

gilles-peskine-arm commented on docs/psa-transition.md:
Changing to “one of the formats”, which is true and doesn't go into needless details.

Hopefully in 3.6 we can replace this whole paragraph by “call `mbedtls_pk_export_to_psa`” or some such.

gilles-peskine-arm commented on docs/psa-transition.md:
This document is about the library as it exists now (or in a couple of places it mentions planned improvements, but it calls them out as such). Part of the work in https://github.com/Mbed-TLS/mbedtls/issues/7760 will be to update this document. It's kind of the whole point of https://github.com/Mbed-TLS/mbedtls/issues/7760 to replace the parts of this document that explain complicated processes by simpler processes.

gilles-peskine-arm commented on docs/psa-transition.md:
I was inconsistent in mentioning vs not mentioning type-specific functions in this section. I think I decided to list them but then stopped midway and forgot. I've reformulated it to list the type-specific functions (for both types) for both signature and verification, but in a separate sentence.

gilles-peskine-arm commented on docs/psa-transition.md:
I agree, and the similar passage about verification does say “a public key or a key pair”. I guess I just forgot here.

mpg replied:
> Minor: I'd prefer writing it as "The key must be a public key (or a keypair) allowing the usage..." as IMO it makes it a bit clearer that "allowing the usage" refers to both, not just the keypair.

Just wanted to check if the reason you didn't add parentheses as suggested is because you disagree or just missed that point? (If you disagree I won't insist, but I do think it makes the sentence less ambiguous.)

gilles-peskine-arm replied:
I copied the wording from the passage about verification. I'm indifferent about the parentheses so I'll add them to both places.

gilles-peskine-arm commented on docs/psa-transition.md:
We don't have a legacy API to deterministically derive a key pair, but that is a thing that application code can do.

gilles-peskine-arm commented on docs/psa-transition.md:
I think it's reasonable to expect that readers who are struggling with code using ECDH (resp. DHM) functions will consult the subsection that is specifically about ECDH (resp. DHM). I'd prefer not to make this bullet list even more complex.

mpg replied:
Note: upon more careful reading of the following subsections, I think none of them mentions using `psa_import_key` or `psa_key_derivation_output_key` ever again. I agree about not making the present list more complex, but if we defer to the following subsections, then they should actually have the material that we're omitting here :)

gilles-peskine-arm replied:
For DH keys, anything other than generating a random key pair and importing/passing the peer's public key is unusual. The subsequent sections cover the usual cases. If you're in not in the usual cases, I think it's reasonable that this document doesn't hold your hand. We can't cover everything. So I don't intend to pursue this point further.

valeriosetti reviewed (commented):
I only found 1 error in the name of a PKCS#12 function. The rest is fine for what concerns with key derivation.

ronald-cron-arm reviewed (commented):
Just one comment about the "General considerations" section.

Ryan-Everett-arm commented:
> Does a green tick in the reviewers table mean "I have done a review" or "I have done a review and given my approval"? Guessing the latter.

I put the first tick in the table to indicate that I had approved my section. In hindsight I probably should have said somewhere what the tick meant.

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `asn1write.h` | `mbedtls_asn1write_` | No change (not a crypto API) |
+| `base64.h` | `mbedtls_base64_` | [PK format support interfaces](#pk-format-support-interfaces) |
+| `bignum.h` | `mbedtls_bignum_` | None (no low-level arithmetic) |
+| `build_info.h` | `MBEDTLS_` | No change (not a crypto API) |
```
```suggestion
| `build_info.h` | N/A | No public APIs (internal support header) |
```
like check_config.h ?

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `chacha20.h` | `mbedtls_chacha20_` | [Symmetric encryption](#symmetric-encryption) |
+| `chachapoly.h` | `mbedtls_chachapoly_` | [Symmetric encryption](#symmetric-encryption) |
+| `check_config.h` | N/A | No public APIs (internal support header) |
+| `cipher.h` | `mbedtls_cipher_` | [Symmetric encryption](#symmetric-encryption) |
```
```suggestion
| `cipher.h` | `mbedtls_cipher_` | [Symmetric encryption](#symmetric-encryption), [Authenticated cipher operations](#authenticated-cipher-operation) |
```
?

gilles-peskine-arm replied:
The AEAD section is a subsection of symmetric encryption. I think it only needs to be called out when there's an emphasis on authenticated encryption.

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `chachapoly.h` | `mbedtls_chachapoly_` | [Symmetric encryption](#symmetric-encryption) |
+| `check_config.h` | N/A | No public APIs (internal support header) |
+| `cipher.h` | `mbedtls_cipher_` | [Symmetric encryption](#symmetric-encryption) |
+| `cmac.h` | `mbedtls_cmac_` | [Hashes and MAC](#hashes-and-mac), [MAC calculation](#mac-calculation) |
```
```suggestion
| `cmac.h` | `mbedtls_cipher_cmac_` | [Hashes and MAC](#hashes-and-mac), [MAC calculation](#mac-calculation) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `cmac.h` | `mbedtls_cmac_` | [Hashes and MAC](#hashes-and-mac), [MAC calculation](#mac-calculation) |
+| `compat-2.x.h` | various | None (transitional APIs) |
+| `config_psa.h` | N/A | No public APIs (internal support header) |
+| `constant_time.h` | `mbedtls_constant_time_` | [Constant-time functions](#constant-time-functions) |
```
```suggestion
| `constant_time.h` | `mbedtls_ct_` | [Constant-time functions](#constant-time-functions) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `ecjpake.h` | `mbedtls_ecjpake_` | [EC-JPAKE](#ec-jpake) |
+| `ecp.h` | `mbedtls_ecp_` | [Asymmetric cryptography](#asymmetric-cryptography) |
+| `entropy.h` | `mbedtls_entropy_` | [Random generation interface](#random-generation-interface), [Entropy sources](#entropy-sources) |
+| `error.h` | `mbedtls_error_` | [Error messages](#error-messages) |
```
```suggestion
| `error.h` | `mbedtls_` | [Error messages](#error-messages) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `gcm.h` | `mbedtls_gcm_` | [Symmetric encryption](#symmetric-encryption), [Authenticated cipher operations](#authenticated-cipher-operations) |
+| `hkdf.h` | `mbedtls_hkdf_` | [HKDF](#hkdf) |
+| `hmac_drbg.h` | `mbedtls_hmac_drbg_` | [Random generation interface](#random-generation-interface), [Deterministic pseudorandom generation](#deterministic-pseudorandom-generation) |
+| `lms.h` | `mbedtls_lms_` | No migration path yet |
```
Probably better to say that this interface will stay.

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `md5.h` | `mbedtls_md5_` | [Hashes and MAC](#hashes-and-mac) |
+| `memory_buffer_alloc.h` | `mbedtls_memory_buffer_alloc_` | No change (not a crypto API) |
+| `net_sockets.h` | `mbedtls_net_` | No change (not a crypto API) |
+| `nist_kw.h` | `mbedtls_nist_kw_` | No migration path yet |
```
Probably better to say that this interface will stay.

gilles-peskine-arm replied:
Actually, no, we're not planning to keep this one in 4.0. But we haven't defined a PSA equivalent yet.

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `pkcs7.h` | `mbedtls_pkcs7_` | No change (not a crypto API) |
+| `pkcs12.h` | `mbedtls_pkcs12_` | [PKCS#12 module](#pkcs12-module) |
+| `platform.h` | `mbedtls_platform_` | No change (not a crypto API) |
+| `platform_time.h` | `mbedtls_platform_time_` | No change (not a crypto API) |
```
```suggestion
| `platform_time.h` | `mbedtls_` | No change (not a crypto API) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `pkcs12.h` | `mbedtls_pkcs12_` | [PKCS#12 module](#pkcs12-module) |
+| `platform.h` | `mbedtls_platform_` | No change (not a crypto API) |
+| `platform_time.h` | `mbedtls_platform_time_` | No change (not a crypto API) |
+| `platform_util.h` | `mbedtls_platform_util_` | No change (not a crypto API) |
```
```suggestion
| `platform_util.h` | `mbedtls_platform_` | No change (not a crypto API) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `platform_util.h` | `mbedtls_platform_util_` | No change (not a crypto API) |
+| `poly1305.h` | `mbedtls_poly1305_` | None (but there is Chacha20-Poly1305 [AEAD](#symmetric-encryption)) |
+| `private_access.h` | N/A | No public APIs (internal support header) |
+| `psa_util.h` | `mbedtls_psa_` | No public APIs (internal support header) |
```
```suggestion
| `psa_util.h` | `mbedtls_psa_`, `psa_` | No public APIs (internal support header) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `sha512.h` | `mbedtls_sha512_` | [Hashes and MAC](#hashes-and-mac) |
+| `ssl.h` | `mbedtls_ssl_` | No change (not a crypto API) |
+| `ssl_cache.h` | `mbedtls_ssl_cache_` | No change (not a crypto API) |
+| `ssl_ciphersuites.h` | `mbedtls_ssl_ciphersuites_` | No change (not a crypto API) |
```
```suggestion
| `ssl_ciphersuites.h` | `mbedtls_ssl_ciphersuite_` | No change (not a crypto API) |
```

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `threading.h` | `mbedtls_threading_` | No change (not a crypto API) |
+| `timing.h` | `mbedtls_timing_` | No change (not a crypto API) |
+| `version.h` | `mbedtls_version_` | No change (not a crypto API) |
+| `x509.h` | `mbedtls_x509` | No change (not a crypto API) |
+| `x509_crl.h` | `mbedtls_x509` | No change (not a crypto API) |
+| `x509_crt.h` | `mbedtls_x509` | No change (not a crypto API) |
+| `x509_csr.h` | `mbedtls_x509` | No change (not a crypto API) |
```
```suggestion
| `x509.h` | `mbedtls_x509_` | No change (not a crypto API) |
| `x509_crl.h` | `mbedtls_x509_crl_` | No change (not a crypto API) |
| `x509_crt.h` | `mbedtls_x509_crt_`, `mbedtls_x509write_crt_` | No change (not a crypto API) |
| `x509_csr.h` | `mbedtls_x509_csr_`, `mbedtls_x509write_csr_` | No change (not a crypto API) |
```

gilles-peskine-arm replied:
Meh, the point of the prefixes is to make it easier to go from function name to relevant section. If there's no relevant section it's all the same.

ronald-cron-arm reviewed (commented):
A few minor comments about the "Summary of API modules" section.

gilles-peskine-arm commented on docs/psa-transition.md:
No. In the next major release, we plan to hide the legacy PKCS5/PKCS12 APIs, but keep calling them from pk. In a subsequent minor release, we'll introduce a PSA API that can import/export encrypted keys (still in the design stage). Also, in a subsequent minor release, we'll change the PBKDF2 code to use the PSA implementation of PBKDF2.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1293 @@
+
+You should probably enable [`MBEDTLS_USE_PSA_CRYPTO`](https://mbed-tls.readthedocs.io/projects/api/en/development/api/file/mbedtls__config_8h/#mbedtls__config_8h_1a70fd7b97d5f11170546583f2095942a6) as well (it is disabled by default). This option causes the PK, X.509 and TLS modules to use PSA crypto under the hood. Some functions that facilitate the transition (for example, to convert between me ...
+
+By default, the PSA crypto API offers a similar set of cryptographic mechanisms as those offered by the legacy API. The PSA crypto API also has its own configuration mechanism; see “[Cryptographic mechanism availability](#cryptographic-mechanism-availability)”.
```
> I would add that it is disabled by default.

That's the very first sentence of the referenced section.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| ------ | --------------- | -------------- |
+| `aes.h` | `mbedtls_aes_` | [Symmetric encryption](#symmetric-encryption) |
+| `aria.h` | `mbedtls_aria_` | [Symmetric encryption](#symmetric-encryption) |
+| `asn1.h` | `mbedtls_asn1_` | No change (not a crypto API) |
+| `asn1write.h` | `mbedtls_asn1write_` | No change (not a crypto API) |
```
I'm removing the hints about 4.0. I was originally unsure when we'd merge this document. Since we now aim to have it in 3.6, I'll focus more on 3.6 and let 4.0 evolution live in GitHub issues.

gilles-peskine-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1298 @@
+| `aes.h` | `mbedtls_aes_` | [Symmetric encryption](#symmetric-encryption) |
+| `aria.h` | `mbedtls_aria_` | [Symmetric encryption](#symmetric-encryption) |
+| `asn1.h` | `mbedtls_asn1_` | No change (not a crypto API) |
+| `asn1write.h` | `mbedtls_asn1write_` | No change (not a crypto API) |
```
“Consistency is the hobgoblin of little minds”, they say — but the full quote is “A foolish consistency is the hobgoblin of little minds”, and I don't think consistency here would be a foolish expectation.

gilles-peskine-arm replied:
I meant that I wish there was more consistency between module names and function prefixes, and I don't think this means I have a little mind. I missed this one when I went and looked for others, I'll fix it.

mpg commented on docs/psa-transition.md:
> This flow can also be used with other protocols

That's more true of the PSA version that the legacy version. IIRC, you can't even use the legacy API directly for TLS 1.3. (Fortunately 1.3 is using the PSA API unconditionally here.) While that is apparent in the details below, I think it would be worth making it more explicit. Perhaps:

"The legacy API offers the following flow for a Diffie-Hellman key agreement in a TLS 1.2 server. Its inputs and outputs use the format prescribed by TLS 1.2. The PSA version of this flow is more generic and can also be used with other protocols, on the side of the party that selects the curve or group and sends its public key first."

mpg commented on docs/psa-transition.md:
Minor: as was observed by another reviewer, I'd prefer "to determine a sufficient size for..." (everywhere).

mpg commented on docs/psa-transition.md:
Note: this assumes ephemeral. However two lines above we has "when doing ephemeral". We should either:
- Only explain ephemeral for the contextless workflow, but then say so in the title/intro and remove the comment two lines above.
- Or fully explain both static and ephemeral. It's not clear to me if that's best done in two separate subsections, or a single one with variants. Wdyt? (Also @tom-daubney-arm as the clarity reviewer.)

tom-daubney-arm replied:
I think the second part of the second point is best, so a variants subsection, provided that it doesn't go out of scope.

gilles-peskine-arm replied:
I'm making this section specific to ephemeral. For static ECDH, I'll write another section once I figure out how to do it with the legacy API.

mpg commented on docs/psa-transition.md:
Typo: descCribed

mpg commented on docs/psa-transition.md:
Well, to me that's a hybrid between ephemeral and static, which some libraries, including OpenSSL I think, used to do instead of real ephemeral, as an optimization: generate a new key only every 100 or so handshakes (or perhaps it was time based?).

What I'd tend to call static is when at step 3 you don't call `psa_generate_key` but instead load the key from somewhere (so, `psa_import_key()` or even `psa_key_derivation_output_key()`). Note that for the TLS 1.2 `ECDH_xxx` key exchanges, your public key is embedded in a certificate, so really you won't generate it when starting your TLS server, but way ahead of that.

Of course not everyone will agree about terminology - my personal view is there's ephemeral, fake ephemeral, and static - but that's not necessarily a problem here, because it's a migration guide, so we can just tell people: if you were using legacy function xxx, then now use PSA function yyy and avoid the question of how to call it.

mpg commented on docs/psa-transition.md:
Minor: perhaps say "restartable/interruptible", or otherwise include the PSA name for it, in order to make improve discoverability.

mpg reviewed (changes requested):
I completed my review of the key agreement part, and I think it may need some restructuring. Right now, only ephemeral (and what I call fake-ephemeral) is described, not what I call static. Regardless of how we name it, and more importantly, how to choose which key creation method to use (generate vs import/derive) is never explained in terms of which legacy functions you were calling.

mpg reviewed (commented):
I've completed my first pass over the Asymmetric crypto section.

valeriosetti commented on docs/psa-transition.md:
```suggestion
The PSA subsystem has an internal random generator. As a consequence, you do not need to instantiate one manually, so most applications using PSA crypto do not need the interfaces from `entropy.h`, `ctr_drbg.h` and `hmac_drbg.h`. See the next sections for remaining use cases for [entropy](#entropy-sources) and [DRBG](#deterministic-pseudorandom-generation).
```

valeriosetti reviewed (commented):
I found only 1 typo (missing header extension) also for the "random generation" part. The rest is fine for me

gilles-peskine-arm commented on docs/psa-transition.md:
> the PSA API is not stable yet

Mostly that, and there was some planned/ongoing changes in the Mbed TLS implementation when I was originally writing that document.

> there are likely fewer users of JPAKE

Also that.

And also that I'm less familiar with ECJPAKE than with most of the rest of the library.

Thanks for listing these, I'll add them to the document. Having a detailed workflow comparison would be good, but I'll defer that to a later pull request (possibly not by me).

gilles-peskine-arm commented:
I believe that I've addressed all the comments except for static ECDH, which I'll rework later.

I've added a section on memory and key management in the “General considerations” chapter. I want to keep that chapter short, but these are significant pervasive differences between the legacy and the PSA API so I think they deserve being shown prominently.

mpg reviewed (commented):
All my feedback has been addressed to my satisfaction - except for static ECDH which you said you'll rework later.

gilles-peskine-arm commented:
@Ryan-Everett-arm Thanks for checking!

1. These are non-crypto-related options, so mostly not relevant for the migration. (The ones relevant to the `error` module are covered in the document though.)
2. Specific symmetric encryption modes are discussed briefly in the introductory paragraph of [“Symmetric encryption”](https://github.com/Mbed-TLS/mbedtls/blob/9f55e8e44267ab76bcae94184716122b6729b517/docs/psa-transition.md#symmetric-encryption): there's no PSA equivalent.
3. `mbedtls_XXX_ALT`, though, does have a migration path, which is PSA drivers. I've [added a section about that](https://github.com/Mbed-TLS/mbedtls/blob/9f55e8e44267ab76bcae94184716122b6729b517/docs/psa-transition.md#alternative-implementations-mbedtls_xxx_alt-options).
4. As with (2): DES is a low-level module with no PSA equivalent.

tom-daubney-arm commented:
Update on my review progress: All my feedback up to this point has been satisfactorily addressed. I have gone through all the sections and I don't think that there are any missing subsections (for the sections I have reviewed). Included below is a list of identifiers (fn names, macros and types) that are not directly searchable in the document. As discussed with @gilles-peskine-arm, much of this is ok, but I leave it here for reference and in case there is the odd one or two in here that ought to be added:
| Topic \\ Header             | `dhm.h`                 | `ecdh.h`                                        | `ecdsa.h`                          | `ecp.h`                        | `pk.h`                                   | `rsa.h`              |
| --------------------------- | ----------------------- | ----------------------------------------------- | ---------------------------------- | ------------------------------ | ---------------------------------------- | -------------------- |
|                             | mbedtls_dhm_parameter   | mbedtls_ecdh_side                               |                                    | mbedtls_ecp_group_id           | mbedtls_pk_rsassa_pss_options            | MBEDTLS_RSA_PKCS_V15 |
|                             | mbedtls_dhm_make_public | mbedtls_ecdh_variant                            |                                    | mbedtls_ecp_curve_type         | mbedtls_pk_load_file (internal function) | MBEDTLS_RSA_PKCS_V21 |
|                             |                         | mbedtls_ecdh_context_mbed (probably not needed) |                                    | mbedtls_ecp_modulus_type       |                                          | MBEDTLS_RSA_SIGN     |
|                             |                         | mbedtls_ecdh_get_params                         |                                    | MBEDTLS_ECP_PF_COMPRESSED      |                                          | MBEDTLS_RSA_CRYPT    |
|                             |                         |                                                 |                                    | mbedtls_ecp_get_type           |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_set_zero           |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_is_zero            |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_point_cmp          |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_point_read_string  |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_tls_read_point     |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_tls_write_point    |                                          |                      |
| Restartable                 |                         |                                                 | mbedtls_ecdsa_restart_ver_ctx      | mbedtls_ecp_restart_mul_ctx    | mbedtls_pk_restart_ctx                   |                      |
|                             |                         |                                                 | mbedtls_ecdsa_restart_det_ctx      | mbedtls_ecp_restart_muladd_ctx |                                          |                      |
|                             |                         |                                                 | mbedtls_ecdsa_sign_det_restartable | mbedtls_ecp_check_budget       |                                          |                      |
|                             |                         |                                                 |                                    | MBEDTLS_ECP_BUDGET             |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_mul_restartable    |                                          |                      |
|                             |                         |                                                 |                                    | mbedtls_ecp_muladd_restartable |                                          |                      |
| Alt support                 |                         |                                                 |                                    |                                | mbedtls_pk_rsa_alt_decrypt_func          |                      |
|                             |                         |                                                 |                                    |                                | mbedtls_pk_rsa_alt_sign_func             |                      |
|                             |                         |                                                 |                                    |                                | mbedtls_pk_rsa_alt_key_len_func          |                      |
| Debug (probably not needed) |                         |                                                 |                                    |                                | mbedtls_pk_debug_type                    |                      |
|                             |                         |                                                 |                                    |                                | mbedtls_pk_debug_item                    |                      |
|                             |                         |                                                 |                                    |                                | MBEDTLS_PK_DEBUG_MAX_ITEMS               |                      |

(The table was supposed to make it clearer, but I am not convinced it does. Let me know if a straight forward list is better.)

gowthamsk-arm commented on docs/psa-transition.md:
For consistency, should we change `MBEDTLS_MD` to `MBEDTLS_MD_xxx`?

gowthamsk-arm commented on docs/psa-transition.md:
Can we also specify `mbedtls_md_hmac` here?

gowthamsk-arm commented on docs/psa-transition.md:
Any particular reason why `MBEDTLS_MD_NONE` is skipped from the list or is it because it's obvious?

gilles-peskine-arm replied:
`NONE` isn't an algorithm, it means “none or unspecified” depending on the API, so I don't think it's useful to mention it.

davidhorstmann-arm commented on docs/psa-transition.md:
It's unclear to me what this is referring to. If it is related to the above note on the PSA API being algorithm-independent, I can't see an interface to AES-CMAC-PRF-128 that is algorithm-specific in the code or the PSA spec.

Thinking about it, this makes more sense if I assume there is a 'not' missing, like so:
```suggestion
The PSA API does not have a direct interface for the AES-CMAC-PRF-128 from RFC 4615 at the time of writing. You can calculate it using the interface to AES-CMAC.
```
Is this what was intended?

I also think that the last sentence would be clearer if it was reworded a bit, so this could change to:
```suggestion
The PSA API does not have a direct interface for the AES-CMAC-PRF-128 from RFC 4615 at the time of writing. You can implement AES-CMAC-PRF-128 simply, using the interface to AES-CMAC.
```

gilles-peskine-arm replied:
Indeed there was a missing “not”.

davidhorstmann-arm commented on docs/psa-transition.md:
This section seems generic, and parts of it are not relevant to MACs specifically. Could it be merged into 'Key management' and replaced with a link to that section?

gilles-peskine-arm replied:
A lot of key management doesn't depend on the key type. But many readers will just want to convert some code that's currently using `mbedtls_md_hmac` or `mbedtls_cmac`, and don't care about generic key management. So I think it's better to have dedicated sections that focus on the topic. For example, this section only discusses the policies that are relevant to MAC operations, and does not go into key formats since MAC keys are just a bunch of bytes.

davidhorstmann-arm reviewed (changes requested):
Looks okay apart from a couple of things

(Have reviewed only Hash and MAC so far)

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1341 @@
+
+If a mechanism is not enabled by `PSA_WANT_xxx`, Mbed TLS will normally not include it. This allows builds that use few features to have a small code size. However, this is not guaranteed: a mechanism that is not explicitly requested can be enabled because it is a dependency of another configuration option, because it is used internally, or because the granularity is not fine enough to distinguis ...
+
+Under the hood, `PSA_WANT_xxx` enables the necessary legacy modules. Note that if a mechanism has a PSA accelerator driver, the corresponding legacy module is typically not needed. Thus applications that use a cryptographic mechanism both through the legacy API and through the PSA API need to explicitly enable both the `PSA_WANT_xxx` symbols and the `MBEDTLS_xxx` symbols.
```
```suggestion
Under the hood, `PSA_WANT_xxx` enables the necessary legacy modules. Note that if a mechanism has a PSA accelerator driver, the corresponding legacy module is typically not needed. Thus applications that use a cryptographic mechanism both through the legacy API and through the PSA API need to explicitly enable both the `PSA_WANT_xxx` symbols and the `MBEDTLS_xxx` symbols in `"mbedtls/mbedtls_config.h"`.
``` ?

gilles-peskine-arm replied:
I'd rather not make that sentence more complex. `MBEDTLS_xxx` doesn't necessarily go in `mbedtls/mbedtls_config.h`: it can go in `MBEDTLS_CONFIG_FILE` or `MBEDTLS_USER_CONFIG_FILE`. And `PSA_WANT_xxx` should go into a PSA config file.

ronald-cron-arm replied:
That makes sense, never mind then. Thanks for the response.

ronald-cron-arm commented on docs/psa-transition.md:
```diff
@@ -0,0 +1,1341 @@
+
+### Optimization options
+
+When PSA Crypto mechanisms are implemented by the built-in code from Mbed TLS, the legacy optimization options (e.g. `MBEDTLS_SHA256_SMALLER`, `MBEDTLS_ECP_WINDOW_SIZE`, etc.) apply to the PSA implementation as well (they invoke the same code under the hood).
```
```suggestion
When PSA Crypto mechanisms are implemented by the built-in code from Mbed TLS, the legacy optimization options (e.g. `MBEDTLS_SHA256_SMALLER`, `MBEDTLS_ECP_WINDOW_SIZE`, etc.) as set in `"mbedtls/mbedtls_config.h"` apply to the PSA implementation as well (they invoke the same code under the hood).
``` ?

ronald-cron-arm reviewed (commented):
Regarding the "Compile-time configuration" section, only two propositions that make things more explicit to me.

ronald-cron-arm reviewed (commented):
One comment about the "Miscellaneous support modules" section.

mpg commented:
For the record, I'm happy with my section except for the part about static (EC)DH. Since it looks like this is unlikely to reveal any new gaps in existing APIs (in my experience, doing static (EC)DH is actually easier with PSA as there are fewer differences between static and ephemeral), I'd be happy to leave that to a follow-up as long as it's tracked.

davidhorstmann-arm commented:
I hereby note for the avoidance of doubt that I'm happy with the 'Hashes and MAC' section and therefore have ticked it.

gowthamsk-arm commented on docs/psa-transition.md:
In case of a CMAC, how do we ensure that an algorithm corresponding to `PSA_WANT_KEY_TYPE_xxx` is available if we are not checking `PSA_WANT_ALG_xxx`? Couple of lines explaining that would be helpful.

gilles-peskine-arm replied:
I don't understand the question. Why would we not check `PSA_WANT_ALG_xxx`?

gowthamsk-arm replied:
Misunderstood the explanation, my bad :)

tom-daubney-arm reviewed (approved):
My sections LGTM, thanks.

ronald-cron-arm reviewed (approved):
LGTM for the sections I've reviewed.

tgonzalezorlandoarm reviewed (approved):
LGTM

gowthamsk-arm reviewed (approved):
LGTM

davidhorstmann-arm reviewed (approved):
I have finished going through symmetric encryption.

LGTM, thanks!

gilles-peskine-arm commented:
@mpg I've [filed an issue](https://github.com/Mbed-TLS/mbedtls/issues/8648) for static ECDH. Is the document ok to merge now as far as you're concerned?

mpg reviewed (commented):
LGTM for asymmetric crypto.
